### ISSUE TITLE
`Osu` struct cannot be cloned

### ISSUE THREAD

**Issue by ev3nvy** [MaxOhn/rosu-v2#5]
# What happened
`Osu` struct does not implement `Clone`.

# What was expected
It was expected that `Osu` struct implements `Clone`, due to the [line in documentation](https://docs.rs/rosu-v2/latest/rosu_v2/struct.Osu.html) saying: 
> The main osu client. Cheap to clone.

# History
Briefly looking through the commit history, the `Clone` trait was never implemented.

# Suggested changes
Currently, we are blocked by `Osu::token_loop_tx` not implementing `Clone`. We could wrap it in an `Arc` but that would break re-authentication (`Drop` gets called for each `Osu` clone, not just the last one).

My suggestion would be to move `Osu::token_loop_tx` into `OsuRef` just like it was before [8cf15a5](https://github.com/MaxOhn/rosu-v2/commit/8cf15a55c028a7b28a6ffba385835603051a86cc). Obviously, this was done for a reason, as `update_worker` takes an `Arc<OsuRef>` meaning `OsuRef` never gets dropped.

To work around this, the following changes must be made:
- make the following methods static: 
  - `OsuRef::handle_status` - 0 usages of `self`
  - `OsuRef::send_request` - 4 usages of `self`
  - `OsuRef::request_token` - 6 usages of `self` (4 after all of the listed methods become static)
- pass `client_id`, `client_secret`, `auth_kind` and `token` to `Token::update_worker` instead of `Arc<OsuRef>`

`OsuRef::handle_status` is straightforward enough.
`OsuRef::send_request` references `OsuRef::http` , `OsuRef::ratelimiter`, `OsuRef::timeout` and `OsuRef::retries`.
`OsuRef::request_token` references `OsuRef::client_id` , `OsuRef::client_secret`, `OsuRef::auth_kind` and `OsuRef::token` (+ everything that `OsuRef::send_request` references).

# Further considerations
Is this even a good idea, or should we just simply remove the "Cheap to clone." part?

Could `OsuRef::ratelimiter`, `OsuRef::timeout` and `OsuRef::retries` (or any combination of the three) be grouped together into a single struct? What would such a struct be named?

Do we just use the default values for `OsuRef::ratelimiter`, `OsuRef::timeout` and `OsuRef::retries` in `OsuRef::request_token` just to reduce the number of required parameters?

Anything else I overlooked?

# Other
- [X] I am willing to work on this if I get a go-ahead.

**Comment by MaxOhn** [comment]
Very nice write-up and great catch, can't believe I've let that incorrect comment dangle in there for so long 😅 

I'm seeing three solutions:
- Wrap `token_loop_tx` in an `Arc<Mutex>` to make it cloneable. Seems reasonable enough but also doesn't sit quite right with me, I'm not a great fan of arcing all fields just so the outer struct can be cloneable.
- Moving the `tx` business closer to `OsuRef` as you mentioned. I'm suspecting that the refactoring necessary for this would come at the cost of code quality which is already arguably not the strongest for the token logic 😅 All-in-all also not a favourite option of mine to solve the issue.
- Simply remove the comment about `Osu` being cloneable. Just as before, this would require the user to wrap their client in an `Arc` themselves if they need to clone it.

Personally, I'm tending towards the third option which is simple and doesn't add complexity for the sake of a new feature that users can already get around by `Arc`ing anyway.

And since `Osu` can't be cloned anyway, I believe the `Arc`s for the `cache` and `metrics` fields could technically be dropped as well or maybe at least replaced by a simple `Box` to not increase the struct size too much. Not quite sure if I'd prefer them boxed or not.

Either way, be my guest to work on this 😃 Otherwise I'll snatch it from you :^)

**Comment by ev3nvy** [comment]
Thanks for the fast response!

> Wrap token_loop_tx in an Arc<Mutex> to make it cloneable. Seems reasonable enough but also doesn't sit quite right with me, I'm not a great fan of arcing all fields just so the outer struct can be cloneable.

This option unfortunately wouldn't work because re-authentication task would stop as soon as one `Osu` clone gets out of scope (unless I'm mistaken). See: 

<details>
<summary>code example</summary>

```rust
#[tokio::main]
async fn main() {
    let osu: Osu = Osu::new(0, "").await.unwrap();
    let osu_cloned = osu.clone();
    drop(osu_cloned);
    // update_worker no longer runs here
}
```
</details>

> Moving the tx business closer to OsuRef as you mentioned. I'm suspecting that the refactoring necessary for this would come at the cost of code quality which is already arguably not the strongest for the token logic 😅 All-in-all also not a favourite option of mine to solve the issue.

This was the option I was leaning towards, but the brought up points make sense + this would be more work.

> And since Osu can't be cloned anyway, I believe the Arcs for the cache and metrics fields could technically be dropped as well or maybe at least replaced by a simple Box to not increase the struct size too much. Not quite sure if I'd prefer them boxed or not.

I am indifferent regarding this, whichever suits you better 😅
